ucsb-cs156/proj-courses
# Staff f22: Add job to update MongoDB for a single quarter / subject area

# Overview

In this PR, we roll out a new feature that allows admins to update a MongoDB database with course information via an asynchronous job.

The intent is for this new feature to open up several new lines of development, especially for the -1 teams on courses:

Especially, this opens up the possibility to develop searches that are not supported by the UCSB Courses API, such as search by course (over multiple quarters) and search by instructor (over multiple quarters.)

It also opens up the possibility to add similar jobs that will:
* load all subject areas in a given quarter.
* load all subject areas over a range of quarters.

# User Story

As an admin
* I can specify a quarter and a subject area and launch a job that loads data for that quarter and subject area into the MongoDB database
* So that developers can work on queries that require such a database

# How to Test / Use this feature

Here's an outline, followed by some animations

1. Configure your app for a fresh MongoDB database; go see the data that is loaded 
2. Login as an admin
3. Go to Admin / Load Subjects to make sure that the subject areas are loaded (this only needs to be done once)
4. Go to Admin / Manage Jobs
5. Specify Update Course Data
6. Specify quarter and subject area
7. Click "Update"
8. Wait for job to finish.  It should report how many courses were loaded or updated
9. See results in MongoDb 


Step 1: See MongoDB data:

This animation shows how to look at and do queries on MongoDB data:

![see-mongodb-data](https://user-images.githubusercontent.com/1119017/203417379-ae587495-b3c5-4afd-87b4-cbb8219124fe.gif)


Steps 2-3: Login as Admin, Load Subjects

![load-subjects](https://user-images.githubusercontent.com/1119017/203416752-1f6838d3-86d5-499d-94d0-17e3567b8751.gif)

Steps: 4-8:  Running the job and waiting for it to finish (in this case, MATH courses in W23):

![load-course-data](https://user-images.githubusercontent.com/1119017/203418212-9a0c075c-03c9-4d0f-b964-16992b45fdbf.gif)

Step 9: See the MATH courses 

The query is:

```
{ 'courseInfo.quarter' : '20231', 'courseInfo.courseId': { $regex: /MATH/ }}
```

![see-math-courses](https://user-images.githubusercontent.com/1119017/203419046-3b50d57b-25be-4b2c-bb96-69d56fd7ed38.gif)
